lib/mapping.js

```diff
@@ -0,0 +1,167 @@
+const {fixWin32Permissions} = require('./item');
+const path = require('path');
+const FileSystem = require('./filesystem');
+const fs = require('fs');
+const bypass = require('./bypass');
+
+const createContext = ({output, options = {}, target}, newContext) =>
+  Object.assign(
+    {
+      // Assign options and set defaults if needed
+      options: {
+        recursive: options.recursive !== false,
+        lazyLoad: options.lazyLoad !== false
+      },
+      output,
+      target
+    },
+    newContext
+  );
+
+function addFile(context, stats, isRoot) {
+  const {output, target} = context;
+  const {lazyLoad} = context.options;
+
+  if (!stats.isFile()) {
+    throw new Error(`${target} is not a valid file!`);
+  }
+
+  const outputPropKey = isRoot ? target : path.win32.basename(target);
+
+  output[outputPropKey] = () => {
+    const content = !lazyLoad ? fs.readFileSync(target) : '';
+    const file = FileSystem.file(Object.assign({}, stats, {content}))();
+
+    if (lazyLoad) {
+      Object.defineProperty(file, '_content', {
+        get() {
+          const res = bypass(() => fs.readFileSync(target));
+          Object.defineProperty(file, '_content', {
+            value: res,
+            writable: true
+          });
+          return res;
+        },
+        set(data) {
+          Object.defineProperty(file, '_content', {
+            value: data,
+            writable: true
+          });
+        },
+        configurable: true
+      });
+    }
+
+    return file;
+  };
+
+  return output[outputPropKey];
+}
+
+function addDir(context, stats, isRoot) {
+  const {target, output} = context;
+  const {recursive} = context.options;
+
+  if (!stats.isDirectory()) {
+    throw new Error(`${target} is not a valid directory!`);
+  }
+
+  stats = Object.assign({}, stats);
+  const outputPropKey = isRoot ? target : path.win32.basename(target);
```

Comment:
Does this work on non-win32 systems?

Reply:
It does, yes. This ensures that basename works with _both_ posix and windows style paths. See [Windows vs. Posix](https://nodejs.org/api/path.html#path_windows_vs_posix).

Example:
```
> path.win32.basename('/a/b/c')
'c'
> path.win32.basename('C:\\windows\\c')
'c'
> path.posix.basename('/c/d')
'd'
> path.posix.basename('C:\\windows\\d')
'C:\\windows\\d'
```

For this use case, I'm not sure whether it would matter, though it might if the user was running from cygwin. Overall, it seemed best practice to use the most universally sound method.

Reply:
Plain `path.basename()` does the right thing for all paths when running on Windows too.

Your code has the advantage to support path `'C:\\windows\\c'` on posix systems. But if users use windows path name while testing code inside linux/mac, they will likely encounter issues in other places even if they have no issue here.

Reply:
I see your code is to load real dir, so it's irrelevant to support `'C:\\windows\\c'` on posix systems.
`path.basename` is enough.

Reply:
*sigh* 

This is simply best practice to standardize behaviour across all environments. As I stated, it may not be necessary, though in cygwin environments it could very likely cause issue. There is, however, zero downside to using it in posix, which is why it's best practice.

Posix result:
```
> path.basename('C:\\d\\e')
'C:\\d\\e'
> path.win32.basename('C:\\d\\e')
'e'
> path.basename('/c/d')
'd'
> path.win32.basename('/c/d')
'd'
>
```

Reply:
I see, I didn't test cygwin env.

Reply:
I really don't think it's a best practice due to there is difference in posix file name and windows file name.

In windows, `\` is not allowed in file name, but it's allowed in posix system.

In posix, `path.win32.basename` fields wrong result.

```⋊> ~/b/bcx-broker-au-frontend on master ◦ node                                                                09:00:24
Welcome to Node.js v12.18.2.
Type ".help" for more information.
> path.basename('/a/b\\c')
'b\\c'
> path.win32.basename('/a/b\\c')
'c'
>

Reply:
> To achieve consistent results when working with Windows file paths on any operating system, use path.win32:
To achieve consistent results when working with POSIX file paths on any operating system, use path.posix:

The doc didn't say use `path.win32` for consistent results with **both Windows and POSIX** file paths. I think what's the argument here.

Reply:
I'm aware of what it says. To be clear, when unsure of whether a path is windows or posix, it is the best practice with native libraries.

I was, however, unaware of the backslash issue. I'll remove the win32 and we'll hope cygwin doesn't take issue.

Reply:
cygwin always got weird issues. Better to test it out if you have the env.

On the other side, I always felt nodejs std lib should normalise the win32/posix path problem, many other languages do that.

Reply:
I agree! Thanks for pointing out the backslash issue.